### DISCUSSION TITLE
"tcp4 0.0.0.0:2379: bind: address already in use"

### DISCUSSION THREAD

**Discussion by lingvisa** [milvus-io/milvus#35476] [Q&A and General discussion]
I am using MacPro. I am following the instruction here to install Milvus:
`https://milvus.io/docs/install_standalone-docker.md`

But got this error message:
```
bash standalone_embed.sh start
Error response from daemon: driver failed programming external connectivity on endpoint milvus-standalone (e13f0c0f0dcd2e6722860c8faec12e1709411e3ad4f00c977dcad4bddec17b0e): listen tcp4 0.0.0.0:2379: bind: address already in use
Error: failed to start containers: milvus-standalone
Start failed.

```
Then execute "sudo lsof -i :2379", but it returns nothing. Then I removed the 'milvus-standalone' from my Docker desktop, and then execute the command again, but received the same message. I then restarted my computer and deleted it again in my Docker containers but the issue remains. 

I am attaching a screenshot  for the container. 
<img width="1439" alt="Screenshot 2024-08-14 at 10 20 27 AM" src="https://github.com/user-attachments/assets/e889e5ec-a2e1-4cf1-ad8e-077508a235a0">

**Comment by lingvisa** [comment]
Then I changed 2379 to 2381 in these two files:
```
cat embedEtcd.yaml 
listen-client-urls: http://0.0.0.0:2381
advertise-client-urls: http://0.0.0.0:2381
quota-backend-bytes: 4294967296
auto-compaction-mode: revision
auto-compaction-retention: '1000'

```

And 2381 from the file standalone_embed.sh as below:
```
run_embed() {
    cat << EOF > embedEtcd.yaml
listen-client-urls: http://0.0.0.0:2381
advertise-client-urls: http://0.0.0.0:2381
quota-backend-bytes: 4294967296
auto-compaction-mode: revision
auto-compaction-retention: '1000'
EOF

    cat << EOF > user.yaml
# Extra config to override default milvus.yaml
EOF

    sudo docker run -d \
        --name milvus-standalone \
        --security-opt seccomp:unconfined \
        -e ETCD_USE_EMBED=true \
        -e ETCD_DATA_DIR=/var/lib/milvus/etcd \
        -e ETCD_CONFIG_PATH=/milvus/configs/embedEtcd.yaml \
        -e COMMON_STORAGETYPE=local \
        -v $(pwd)/volumes/milvus:/var/lib/milvus \
        -v $(pwd)/embedEtcd.yaml:/milvus/configs/embedEtcd.yaml \
        -v $(pwd)/user.yaml:/milvus/configs/user.yaml \
        -p 19530:19530 \
        -p 9091:9091 \
        -p 2381:2381 \
        --health-cmd="curl -f http://localhost:9091/healthz" \
        --health-interval=30s \
        --health-start-period=90s \
        --health-timeout=20s \
        --health-retries=3 \
        milvusdb/milvus:v2.4.5 \
        milvus run standalone  1> /dev/null
}

```

But it reports the same error, still 2379 in use.

**Comment by xiaofan-luan** [comment]
try sudo netstat -tuln | grep :2379 to see which process used 2379 port.
you can either kill that process or change the port of milvus port.
I would recommend the prior

**Reply by lingvisa** [reply]
```
 % sudo netstat -tuln | grep :2379

```
As you can see, it didn't return anything. Instead I can see it in the containers in Docker, as you can see from the screenshot.

**Comment by lingvisa** [comment]
@xiaofan-luan I can start the container correctly by using 'docker compose -f milvus-standalone-docker-compose.yml up -d'. That's the same thing as 'bash standalone_embed.sh start'. Is that right? If that's the case, my problem is solved.

**Comment by yhmo** [comment]
@lingvisa 
The **embedEtcd.yaml** defines the port that the embed Etcd server is listening to.
But the milvus server has its configuration file **milvus.yaml** to configure the Etcd server address. If you take a look the default milvus.yaml, you will see the etcd configuration for milvus:
https://github.com/milvus-io/milvus/blob/35cc7dde6380777842d06d5a0f3858d3f29c17cf/configs/milvus.yaml#L22

By default, milvus will try to connect etcd with port 2379:
```
etcd:
  endpoints: localhost:2379
```

The **user.yaml** allow you to change this value in **milvus.yaml**.

Open the file **standalone_embed.sh**, we can see the first time it runs into the `run_embed()` function to reset the **embedEtcd.yaml** and **user.yaml**.
So, we can modify the **standalone_embed.sh** to be:
```
run_embed() {
    cat << EOF > embedEtcd.yaml
listen-client-urls: http://0.0.0.0:2381
advertise-client-urls: http://0.0.0.0:2381
quota-backend-bytes: 4294967296
auto-compaction-mode: revision
auto-compaction-retention: '1000'
EOF

    cat << EOF > user.yaml
# Extra config to override default milvus.yaml
etcd:
  endpoints: localhost:2381
  
EOF

    sudo docker run -d \
        --name milvus-standalone \
        --security-opt seccomp:unconfined \
        -e ETCD_USE_EMBED=true \
        -e ETCD_DATA_DIR=/var/lib/milvus/etcd \
        -e ETCD_CONFIG_PATH=/milvus/configs/embedEtcd.yaml \
        -e COMMON_STORAGETYPE=local \
        -v $(pwd)/volumes/milvus:/var/lib/milvus \
        -v $(pwd)/embedEtcd.yaml:/milvus/configs/embedEtcd.yaml \
        -v $(pwd)/user.yaml:/milvus/configs/user.yaml \
        -p 19530:19530 \
        -p 9091:9091 \
        -p 2381:2381 \
        --health-cmd="curl -f http://localhost:9091/healthz" \
        --health-interval=30s \
        --health-start-period=90s \
        --health-timeout=20s \
        --health-retries=3 \
        milvusdb/milvus:v2.4.6 \
        milvus run standalone  1> /dev/null
}
```

If there is already a milvus-standalone container, this function will not be hit. So, we delete the milvus-standalone container firstly by this command:
`bash standalone_embed.sh delete`

Then start a new container:
`bash standalone_embed.sh start`

Now we export the log of the new container:
`docker logs [container id] > 1.log`

We will see this line in the log that means the port of the Etcd has been changed successfully:
`[2024/08/15 03:01:39.118 +00:00] [INFO] [etcd/etcd_util.go:49] ["create etcd client"] [useEmbedEtcd=true] [useSSL=false] [endpoints="[localhost:2381]"] [minVersion=1.3]`

**Reply by lingvisa** [reply]
Where did you modify the file? I can't see it.

**Reply by yhmo** [reply]
In the standalone_embed.sh, in the run_embed() function.
Just change the text content below the two lines:
cat << EOF > embedEtcd.yaml
cat << EOF > user.yaml

```
run_embed() {
    cat << EOF > embedEtcd.yaml
listen-client-urls: http://0.0.0.0:2381
advertise-client-urls: http://0.0.0.0:2381
quota-backend-bytes: 4294967296
auto-compaction-mode: revision
auto-compaction-retention: '1000'
EOF

    cat << EOF > user.yaml
# Extra config to override default milvus.yaml
etcd:
  endpoints: localhost:2381
EOF
```

**Reply by lingvisa** [reply]
This is my version of the function:
```
run_embed() {
    cat << EOF > embedEtcd.yaml
listen-client-urls: http://0.0.0.0:2379/
advertise-client-urls: http://0.0.0.0:2379/
quota-backend-bytes: 4294967296
auto-compaction-mode: revision
auto-compaction-retention: '1000'
EOF

    cat << EOF > user.yaml
# Extra config to override default milvus.yaml
EOF

    sudo docker run -d \
        --name milvus-standalone \
        --security-opt seccomp:unconfined \
        -e ETCD_USE_EMBED=true \
        -e ETCD_DATA_DIR=/var/lib/milvus/etcd \
        -e ETCD_CONFIG_PATH=/milvus/configs/embedEtcd.yaml \
        -e COMMON_STORAGETYPE=local \
        -v $(pwd)/volumes/milvus:/var/lib/milvus \
        -v $(pwd)/embedEtcd.yaml:/milvus/configs/embedEtcd.yaml \
        -v $(pwd)/user.yaml:/milvus/configs/user.yaml \
        -p 19530:19530 \
        -p 9091:9091 \
        -p 2379:2379 \
        --health-cmd="curl -f http://localhost:9091/healthz" \
        --health-interval=30s \
        --health-start-period=90s \
        --health-timeout=20s \
        --health-retries=3 \
        milvusdb/milvus:v2.4.5 \
        milvus run standalone  1> /dev/null
}
```
Do you mean insert "etcd:
  endpoints: localhost:2379"  to it?

**Reply by yhmo** [reply]
Yes, replace 2379 with another port number. And append these two lines under the cat << EOF > user.yaml
```
etcd:
  endpoints: localhost:2381
```
My function replaced 2379 to be 2381:
```
run_embed() {
    cat << EOF > embedEtcd.yaml
listen-client-urls: http://0.0.0.0:2381
advertise-client-urls: http://0.0.0.0:2381
quota-backend-bytes: 4294967296
auto-compaction-mode: revision
auto-compaction-retention: '1000'
EOF

    cat << EOF > user.yaml
# Extra config to override default milvus.yaml
etcd:
  endpoints: localhost:2381
  
EOF

    sudo docker run -d \
        --name milvus-standalone \
        --security-opt seccomp:unconfined \
        -e ETCD_USE_EMBED=true \
        -e ETCD_DATA_DIR=/var/lib/milvus/etcd \
        -e ETCD_CONFIG_PATH=/milvus/configs/embedEtcd.yaml \
        -e COMMON_STORAGETYPE=local \
        -v $(pwd)/volumes/milvus:/var/lib/milvus \
        -v $(pwd)/embedEtcd.yaml:/milvus/configs/embedEtcd.yaml \
        -v $(pwd)/user.yaml:/milvus/configs/user.yaml \
        -p 19530:19530 \
        -p 9091:9091 \
        -p 2381:2381 \
        --health-cmd="curl -f http://localhost:9091/healthz" \
        --health-interval=30s \
        --health-start-period=90s \
        --health-timeout=20s \
        --health-retries=3 \
        milvusdb/milvus:v2.4.6 \
        milvus run standalone  1> /dev/null
}
```

Then recreate the container by the two command:
```
bash standalone_embed.sh delete
bash standalone_embed.sh start
```

**Comment by yhmo** [comment]
@lingvisa 
A possible reason for you got the error "tcp4 0.0.0.0:2379: bind: address already in use".

Because you have installed milvus by DEB/RPM before, there is a milvus service running in the background.
https://milvus.io/docs/v2.3.x/install_standalone-aptyum.md

If you use a command `sudo systemctl status milvus` to check, you will see the background service. The background server launched an etcd service so that port 2379 is occupied. Actually, the port 19530 is also occupied.

So, if you want to install a new milvus by docker-compose, you need to stop the background milvus service.
`sudo systemctl stop milvus`
